File: src/fragile/fractalai/videogames/cloning.py
"""Fractal gas cloning with uniform companion selection and distance-based fitness."""

from dataclasses import dataclass

import torch
from torch import Tensor

from fragile.fractalai.fractalai import (
    asymmetric_rescale,
    l2_norm,
    random_alive_compas,
)


@dataclass
class FractalCloningOperator:
    """Cloning operator using uniform companion selection and distance-based fitness.

    Uses uniform random companion pairing (not distance-based) to calculate fitness
    via L2 distances on observations, then decides cloning probabilistically based
    on virtual reward comparisons.

    Args:
        dist_coef: Distance coefficient for fitness (default 1.0)
        reward_coef: Reward coefficient for fitness (default 1.0)
        use_cumulative_reward: Use cumulative rewards for fitness (default False, uses step rewards)
        device: Device for tensor operations (cpu or cuda)
        eps: Small constant to avoid division by zero (default 1e-8)
    """

    dist_coef: float = 1.0
    reward_coef: float = 1.0
    use_cumulative_reward: bool = False
    device: str = "cpu"
    eps: float = 1e-8

    def __post_init__(self):
        """Initialize tracking attributes."""
        self.last_companions = None
        self.last_will_clone = None
        self.last_virtual_rewards = None
        self.last_fitness_companions = None
        self.last_clone_companions = None

    def calculate_fitness(
        self,
        observations: Tensor,
        cumulative_rewards: Tensor,
        step_rewards: Tensor,
        alive: Tensor,
    ) -> tuple[Tensor, Tensor]:
        """Calculate virtual rewards using L2 distance and rewards.

        Args:
            observations: Walker observations [N, ...], flattened for distance computation
            cumulative_rewards: Total accumulated rewards [N]
            step_rewards: Rewards from last step only [N]
            alive: Boolean mask [N], True for alive walkers

        Returns:
            virtual_rewards: Virtual rewards [N]
            companions: Companion indices [N]
        """
        N = observations.shape[0]

        # Select uniform random companions (alive only)
        companions = random_alive_compas(~alive, observations)

        # Compute L2 distance between walker and companion observations
        obs_flat = observations.reshape(N, -1)
        comp_obs_flat = obs_flat[companions]
        distances = l2_norm(obs_flat, comp_obs_flat)

        # Select reward signal based on parameter
        reward_signal = cumulative_rewards if self.use_cumulative_reward else step_rewards

        # Normalize distances and rewards
        distance_norm = asymmetric_rescale(distances)
        reward_norm = asymmetric_rescale(reward_signal)

        # Compute virtual rewards: distance^dist_coef * reward^reward_coef
        virtual_rewards = distance_norm**self.dist_coef * reward_norm**self.reward_coef

        # Store for diagnostics
        self.last_fitness_companions = companions
        self.last_virtual_rewards = virtual_rewards

        return virtual_rewards, companions

    def decide_cloning(self, virtual_rewards: Tensor, alive: Tensor) -> tuple[Tensor, Tensor]:
        """Decide which walkers should clone based on virtual rewards.

        Args:
            virtual_rewards: Virtual rewards [N]
            alive: Boolean mask [N], True for alive walkers

        Returns:
            companions: Companion indices for cloning [N]
            will_clone: Boolean mask [N], True for walkers that will clone
        """
        # Select new uniform random companions for cloning decision
        companions = random_alive_compas(~alive, virtual_rewards)

        # Compute clone probability: (companion_vr - walker_vr) / walker_vr
        vr = virtual_rewards.flatten()
        vr_comp = vr[companions]
        clone_probs = (vr_comp - vr) / torch.where(
            vr > self.eps,
            vr,
            torch.tensor(self.eps, device=vr.device, dtype=vr.dtype),
        )

        # Sample cloning decisions probabilistically
        will_clone = clone_probs > torch.rand_like(clone_probs)

        # Dead walkers always clone
        will_clone = will_clone | ~alive

        # Store for diagnostics
        self.last_clone_companions = companions
        self.last_will_clone = will_clone

        return companions, will_clone

    def apply(
        self,
        observations: Tensor,
        cumulative_rewards: Tensor,
        step_rewards: Tensor,
        alive: Tensor,
    ) -> tuple[Tensor, Tensor, Tensor]:
        """Combined fitness calculation and cloning decision.

        Args:
            observations: Walker observations [N, ...]
            cumulative_rewards: Total accumulated rewards [N]
            step_rewards: Rewards from last step only [N]
            alive: Boolean mask [N], True for alive walkers

        Returns:
            virtual_rewards: Virtual rewards [N]
            companions: Companion indices for cloning [N]
            will_clone: Boolean mask [N], True for walkers that will clone
        """
        # Calculate fitness
        virtual_rewards, _fitness_companions = self.calculate_fitness(
            observations, cumulative_rewards, step_rewards, alive
        )

        # Decide cloning
        clone_companions, will_clone = self.decide_cloning(virtual_rewards, alive)

        # Store combined companions (use clone companions as the final result)
        self.last_companions = clone_companions

        return virtual_rewards, clone_companions, will_clone


def clone_walker_data(data: Tensor, companions: Tensor, will_clone: Tensor) -> Tensor:
    """Clone walker data arrays based on cloning decisions.

    Args:
        data: Walker data array [N, ...] or [N]
        companions: Companion indices [N]
        will_clone: Boolean mask [N], True for walkers that will clone

    Returns:
        Cloned data array with same shape as input
    """
    cloned_data = data.clone()
    cloned_data[will_clone] = data[companions[will_clone]]
    return cloned_data

File: fractal-gas-web/src/cloning.cpp
#include "cloning.hpp"

#include <cmath>

#include "tensor_ops.hpp"

namespace fg {

std::vector<float> FractalCloningOperator::fitness_with_companions(
    const std::vector<float>& observations, int32_t n, int32_t obs_dim,
    const std::vector<float>& cumulative_rewards,
    const std::vector<float>& step_rewards,
    const std::vector<int32_t>& companions) const {
  const std::vector<float> distances =
      l2_norm_companions(observations, companions, n, obs_dim, pool);
  const std::vector<float>& reward_signal =
      use_cumulative_reward ? cumulative_rewards : step_rewards;

  const std::vector<float> distance_norm = asymmetric_rescale(distances);
  const std::vector<float> reward_norm = asymmetric_rescale(reward_signal);

  std::vector<float> virtual_rewards(static_cast<size_t>(n));
  for (size_t i = 0; i < static_cast<size_t>(n); ++i) {
    virtual_rewards[i] = std::pow(distance_norm[i], dist_coef) *
                         std::pow(reward_norm[i], reward_coef);
  }
  return virtual_rewards;
}

std::vector<float> FractalCloningOperator::clone_probs_with_companions(
    const std::vector<float>& virtual_rewards,
    const std::vector<int32_t>& companions) const {
  const size_t n = virtual_rewards.size();
  std::vector<float> probs(n);
  for (size_t i = 0; i < n; ++i) {
    const float vr = virtual_rewards[i];
    const float vr_comp = virtual_rewards[static_cast<size_t>(companions[i])];
    const float denom = vr > eps ? vr : eps;
    probs[i] = (vr_comp - vr) / denom;
  }
  return probs;
}

std::vector<uint8_t> FractalCloningOperator::decide_with_uniforms(
    const std::vector<float>& clone_probs, const std::vector<float>& uniforms,
    const std::vector<uint8_t>& alive) const {
  const size_t n = clone_probs.size();
  std::vector<uint8_t> will_clone(n);
  for (size_t i = 0; i < n; ++i) {
    const bool clones = clone_probs[i] > uniforms[i];
    will_clone[i] = (clones || !alive[i]) ? 1 : 0;  // dead walkers always clone
  }
  return will_clone;
}

std::vector<int32_t> FractalCloningOperator::sample_companions(
    const std::vector<uint8_t>& alive, Rng& rng) const {
  return random_alive_compas(alive, rng);
}

std::vector<float> FractalCloningOperator::sample_uniforms(int32_t n,
                                                           Rng& rng) const {
  std::vector<float> u(static_cast<size_t>(n));
  for (auto& v : u) v = rng.uniform01();
  return u;
}

std::pair<std::vector<float>, std::vector<int32_t>>
FractalCloningOperator::calculate_fitness(
    const std::vector<float>& observations, int32_t n, int32_t obs_dim,
    const std::vector<float>& cumulative_rewards,
    const std::vector<float>& step_rewards, const std::vector<uint8_t>& alive,
    Rng& rng) const {
  std::vector<int32_t> companions = sample_companions(alive, rng);
  std::vector<float> virtual_rewards = fitness_with_companions(
      observations, n, obs_dim, cumulative_rewards, step_rewards, companions);
  return {std::move(virtual_rewards), std::move(companions)};
}

std::pair<std::vector<int32_t>, std::vector<uint8_t>>
FractalCloningOperator::decide_cloning(
    const std::vector<float>& virtual_rewards,
    const std::vector<uint8_t>& alive, Rng& rng) const {
  const auto n = static_cast<int32_t>(virtual_rewards.size());
  std::vector<int32_t> companions = sample_companions(alive, rng);
  const std::vector<float> probs =
      clone_probs_with_companions(virtual_rewards, companions);
  const std::vector<float> uniforms = sample_uniforms(n, rng);
  std::vector<uint8_t> will_clone = decide_with_uniforms(probs, uniforms, alive);
  return {std::move(companions), std::move(will_clone)};
}

}  // namespace fg
